### ISSUE TITLE
Conversation class proposal

### ISSUE THREAD

**Issue by inventor2525** [AbanteAI/archive-old-cli-mentat#111]
@biobootloader @waydegg @PCSwingle

I see a lot of indecisiveness around the conversation class with a lot of things that seem out of scope for it.

May I propose a structure with better separation of concerns with a bit more forethought (possibly a future dependency)?

Note, I'm thinking ahead here to fully logged model agnostic multi agent conversations (similar to [ChatDev](https://github.com/OpenBMB/ChatDev)), with model training in mind where manual & automated training data augmentation can occur with traceability to obey TOS of any model used. -- I know, that's a bit much, but I think that will ultimately be needed for any such AI pair programming system to scale.

(Mind you, I want to really emphasis fine tuning here, because even if you favor OpenAI models, likely GPT4 will be fine tunable in time as GPT3.5 now is.)

I've been prototyping this in [AbstractAI](https://github.com/inventor2525/AbstractAI), and am currently re-working it locally for readability, simplicity and easier serialization in sql lite, json, and transport & syncing via flask... But when I saw the number of changes to the conversation class which I'm currently using to log my mentat conversations on my instance, I wondered if a discussion wasn't warranted.

```python
@dataclass
class MessageSource:
	pass
	
@dataclass
class Message:
	content: str
	
	creation_time: datetime = field(default_factory=datetime.now)
	
	prev_message: "Message" = None
	conversation: "Conversation" = None
	
	source: MessageSource = None
	
	_children: List["Message"] = internal_field(default_factory=list)

@dataclass
class MessageSequence:
	conversation: "Conversation"
	messages: List[Message] = field(default_factory=list)
	
	def add_message(self, message:Message):
		message.prev_message = None if len(self.messages) == 0 else self.messages[-1]
		self.messages.append(message)
	
@dataclass
class Conversation:
	name: str
	description: str
	
	creation_time: datetime = field(default_factory=datetime.now)
	
	message_sequence:MessageSequence = None
	
	_all_messages: List[Message] = internal_field(default_factory=list)
	_root_messages: List[Message] = internal_field(default_factory=list)
	
	def __post_init__(self, message_sequence:MessageSequence = None):
		if message_sequence is None:
			self.message_sequence = MessageSequence(self)
			
	def add_message(self, message:Message):
		self._all_messages.append(message)
		self.message_sequence.add_message(message)
		if message.prev_message is None:
			self._root_messages.append(message)
		else:
			message.prev_message._children.append(message)

@dataclass
class EditSource(MessageSource):
	original: Message
	new: Message
	new_message_source: MessageSource
	
@dataclass
class ModelSource(MessageSource):
	model_name: str
	model_parameters: dict
	message_sequence: MessageSequence

@dataclass
class UserSource(MessageSource):
	user_name: str = None
```

**Comment by biobootloader** [comment]
Thanks for the suggestion and writeup! Once we finish changes to support generic interfaces and make things async it'll be good to look more into this.

**Comment by waydegg** [comment]
Thank you for the writeup! 

We don't really have any structure in place for handling multi-agent workflows, just one conversation/chat at a time. We're pushing some large changes in the next day or so that will move towards concurrent LLM streams though (basically making a lot of synchronous code async). 

I haven't looked at how [ChatDev](https://github.com/OpenBMB/ChatDev) does it, but for Mentat I'm thinking it would make sense to have some layer on top of the `Conversation` (e.g. an `Agent`) that basically manages multiple conversations.

**Comment by inventor2525** [comment]
Agreed. I think that's more where [AutoGen](https://github.com/microsoft/autogen) is coming in too. I haven't looked at their implementation much yet (life, hard), but Agent is just a compute graph of conversations. Conversations being the edges, LLMs or their configurations being the vertices.

Seems a lot of people are getting on board re-inventing the wheel again there, but everything still applies.

Sub graphs for instance. An agent can be thought of as many smaller agents.

One thing I ?think? people are kinda missing this time still is the f(x). (the non-linearization/'activation function' in NNs). In conversations, that's the summarization. That's the the message ordering, file selection, diff formatting, etc. -- But it appears to me we're re-implementing it again in fixed code. Coomon to start with, it's easier, but: What ever summarization there is needs to be a specific 'model' with high separation of concerns like anything else, so that it too can be replaced latter with something that can be trained.

Inevitably for the code base summarization problem for instance (which I suspect will always be a problem, given that... We clearly do not hold entire code bases in our heads in a moment and yet work with code fairly well), I suspect a combination of embedding and a sort of exploration, gathering, and summarizing agent are going to be needed. One where it can be asked: what is most pertinent to what I am currently wanting to do?

Another might be to verify any code written doesn't re-invent something that couldn't be fit into the context.

For such agents to be trainable though, you need traceability and easy introspection at the start.

I would put forth though that you do have 2 conversations: what the user sees and what the model sees.
